File: packages/hpack-html-py/hpack_html/decoder.py
"""Packet decoder and unpack API."""

from __future__ import annotations

import gzip
import zlib
from dataclasses import dataclass, field
from typing import Optional

from .format import (
    MAGIC, MAGIC_SIZE, VERSION, CRC32_SIZE, MIN_PACKET_SIZE,
    FLAG_MINIFIED, FLAG_CHECKSUM, FLAG_COMPRESSION_MASK, FLAG_COMPRESSION_SHIFT,
    COMPRESSION_GZIP, FIELD_URL, FIELD_ETAG, FIELD_SIGNATURE,
    FIELD_CONTENT_TYPE, FIELD_TIMESTAMP, FIELD_ENCODING, FIELD_CUSTOM_START,
)
from .varint import decode_varint


# Errors
class InvalidMagicError(Exception):
    pass

class UnsupportedVersionError(Exception):
    def __init__(self, version: int):
        super().__init__(f"Unsupported version: {version}")
        self.version = version

class TruncatedPacketError(Exception):
    pass

class ChecksumMismatchError(Exception):
    def __init__(self, expected: int, actual: int):
        super().__init__(f"CRC32 mismatch: expected 0x{expected:08X}, got 0x{actual:08X}")
        self.expected = expected
        self.actual = actual

class DecompressionError(Exception):
    pass


@dataclass
class UnpackResult:
    url: str = ""
    etag: Optional[str] = None
    signature: Optional[str] = None
    content_type: Optional[str] = None
    timestamp: Optional[int] = None
    encoding: Optional[str] = None
    custom: Optional[dict[str, str]] = None
    html: str = ""
    version: int = 0
    minified: bool = False
    compression_algorithm: str = "deflate"
    checksum_valid: Optional[bool] = None


@dataclass
class UnpackOptions:
    verify_checksum: bool = True
    headers_only: bool = False


def _decode_packet(data: bytes | bytearray) -> dict:
    if len(data) < MIN_PACKET_SIZE:
        raise TruncatedPacketError(f"Packet too small: {len(data)} bytes")

    offset = 0

    # Magic
    if data[:MAGIC_SIZE] != MAGIC:
        raise InvalidMagicError()
    offset += MAGIC_SIZE

    # Version
    ver = data[offset]; offset += 1
    if ver != VERSION:
        raise UnsupportedVersionError(ver)

    # Flags
    flags_byte = data[offset]; offset += 1
    minified = bool(flags_byte & FLAG_MINIFIED)
    has_checksum = bool(flags_byte & FLAG_CHECKSUM)
    compression = (flags_byte & FLAG_COMPRESSION_MASK) >> FLAG_COMPRESSION_SHIFT

    # Header length
    header_len, n = decode_varint(data, offset); offset += n
    header_end = offset + header_len
    if header_end > len(data):
        raise TruncatedPacketError("Header section extends beyond packet")

    # Field count
    field_count, n = decode_varint(data, offset); offset += n

    # Fields
    fields: list[dict] = []
    for i in range(field_count):
        if offset >= header_end:
            raise TruncatedPacketError(f"Header field {i} extends beyond header")
        field_type = data[offset]; offset += 1

        if field_type >= FIELD_CUSTOM_START:
            if offset >= header_end:
                raise TruncatedPacketError(f"Custom field {i} name length missing")
            name_len = data[offset]; offset += 1
            if offset + name_len > header_end:
                raise TruncatedPacketError(f"Custom field {i} name truncated")
            name = data[offset:offset + name_len].decode("utf-8"); offset += name_len

            value_len, n = decode_varint(data, offset); offset += n
            if offset + value_len > header_end:
                raise TruncatedPacketError(f"Custom field {i} value truncated")
            value = data[offset:offset + value_len].decode("utf-8"); offset += value_len

            fields.append({"type": field_type, "name": name, "value": value})
        else:
            value_len, n = decode_varint(data, offset); offset += n
            if offset + value_len > header_end:
                raise TruncatedPacketError(f"Field {i} value truncated")
            value = data[offset:offset + value_len].decode("utf-8"); offset += value_len

            fields.append({"type": field_type, "value": value})

    offset = header_end

    # Body
    if offset >= len(data):
        raise TruncatedPacketError("Body section missing")

    uncompressed_len, n = decode_varint(data, offset); offset += n
    crc_size = CRC32_SIZE if has_checksum else 0
    body_end = len(data) - crc_size

    if offset > body_end:
        raise TruncatedPacketError("Compressed body truncated")

    compressed_body = data[offset:body_end]

    crc32_value = None
    if has_checksum:
        crc_off = body_end
        crc32_value = (
            data[crc_off]
            | (data[crc_off + 1] << 8)
            | (data[crc_off + 2] << 16)
            | (data[crc_off + 3] << 24)
        ) & 0xFFFFFFFF

    return {
        "version": ver,
        "minified": minified,
        "has_checksum": has_checksum,
        "compression": compression,
        "fields": fields,
        "compressed_body": compressed_body,
        "uncompressed_len": uncompressed_len,
        "crc32": crc32_value,
    }


def _extract_fields(fields: list[dict]) -> dict:
    result: dict = {"url": "", "etag": None, "signature": None, "content_type": None,
                    "timestamp": None, "encoding": None, "custom": None}
    for f in fields:
        ft = f["type"]
        if ft == FIELD_URL:
            result["url"] = f["value"]
        elif ft == FIELD_ETAG:
            result["etag"] = f["value"]
        elif ft == FIELD_SIGNATURE:
            result["signature"] = f["value"]
        elif ft == FIELD_CONTENT_TYPE:
            result["content_type"] = f["value"]
        elif ft == FIELD_TIMESTAMP:
            try:
                result["timestamp"] = int(f["value"])
            except ValueError:
                pass
        elif ft == FIELD_ENCODING:
            result["encoding"] = f["value"]
        elif ft >= FIELD_CUSTOM_START and "name" in f:
            if result["custom"] is None:
                result["custom"] = {}
            result["custom"][f["name"]] = f["value"]
    return result


def unpack(data: bytes | bytearray, options: UnpackOptions | None = None) -> UnpackResult:
    """Unpack an .hpack binary packet."""
    opts = options or UnpackOptions()
    packet = _decode_packet(data)
    meta = _extract_fields(packet["fields"])

    comp_algo = "gzip" if packet["compression"] == COMPRESSION_GZIP else "deflate"

    if opts.headers_only:
        return UnpackResult(
            **meta, html="", version=packet["version"],
            minified=packet["minified"], compression_algorithm=comp_algo,
        )

    # Decompress
    try:
        if packet["compression"] == COMPRESSION_GZIP:
            decompressed = gzip.decompress(packet["compressed_body"])
        else:
            decompressed = zlib.decompress(packet["compressed_body"], -15)
    except Exception as e:
        raise DecompressionError(str(e)) from e

    # CRC32
    checksum_valid = None
    if packet["has_checksum"] and packet["crc32"] is not None:
        actual = zlib.crc32(decompressed) & 0xFFFFFFFF
        if opts.verify_checksum and actual != packet["crc32"]:
            raise ChecksumMismatchError(packet["crc32"], actual)
        checksum_valid = actual == packet["crc32"]

    html = decompressed.decode("utf-8")

    return UnpackResult(
        **meta, html=html, version=packet["version"],
        minified=packet["minified"], compression_algorithm=comp_algo,
        checksum_valid=checksum_valid,
    )


def unpack_headers(data: bytes | bytearray) -> UnpackResult:
    """Read only headers without decompressing the body."""
    return unpack(data, UnpackOptions(headers_only=True))

File: packages/hpack-html-rs/src/decoder.rs
use crate::varint::decode_varint;
use crate::Error;
use std::collections::HashMap;

// Format constants (must match TypeScript)
const MAGIC: [u8; 4] = [0x89, 0x48, 0x50, 0x4B];
const VERSION: u8 = 0x01;
const FLAG_MINIFIED: u8 = 1 << 0;
const FLAG_CHECKSUM: u8 = 1 << 1;
const FLAG_COMPRESSION_MASK: u8 = 0b00001100;
const FLAG_COMPRESSION_SHIFT: u8 = 2;
const CRC32_SIZE: usize = 4;
const MIN_PACKET_SIZE: usize = 7; // magic(4) + version(1) + flags(1) + headerLen(1)
const FIELD_CUSTOM_START: u8 = 0x10;

// Known field type IDs
const FIELD_URL: u8 = 0x01;
const FIELD_ETAG: u8 = 0x02;
const FIELD_SIGNATURE: u8 = 0x03;
const FIELD_CONTENT_TYPE: u8 = 0x04;
const FIELD_TIMESTAMP: u8 = 0x05;
const FIELD_ENCODING: u8 = 0x06;

/// Compression algorithm.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CompressionAlgorithm {
    Deflate,
    Gzip,
}

/// Result of unpacking an .hpack packet.
#[derive(Debug)]
pub struct UnpackResult {
    pub url: String,
    pub etag: Option<String>,
    pub signature: Option<String>,
    pub content_type: Option<String>,
    pub timestamp: Option<u64>,
    pub encoding: Option<String>,
    pub custom: HashMap<String, String>,
    pub html: String,
    pub version: u8,
    pub minified: bool,
    pub compression: CompressionAlgorithm,
    pub checksum_valid: Option<bool>,
}

/// Options for unpacking.
pub struct UnpackOptions {
    pub verify_checksum: bool,
    pub headers_only: bool,
}

impl Default for UnpackOptions {
    fn default() -> Self {
        Self {
            verify_checksum: true,
            headers_only: false,
        }
    }
}

/// Internal decoded header field.
struct HeaderField {
    field_type: u8,
    name: Option<String>,
    value: String,
}

/// Decode the packet header and body structure.
#[allow(clippy::type_complexity)]
fn decode_packet(
    data: &[u8],
) -> Result<
    (
        u8,                   // version
        bool,                 // minified
        bool,                 // has_checksum
        CompressionAlgorithm, // compression
        Vec<HeaderField>,     // fields
        &[u8],                // compressed_body
        usize,                // uncompressed_length
        Option<u32>,          // crc32
    ),
    Error,
> {
    if data.len() < MIN_PACKET_SIZE {
        return Err(Error::TruncatedPacket(format!(
            "Packet too small: {} bytes (minimum: {})",
            data.len(),
            MIN_PACKET_SIZE
        )));
    }

    let mut offset = 0;

    // Magic
    if data[0..4] != MAGIC {
        return Err(Error::InvalidMagic);
    }
    offset += 4;

    // Version
    let version = data[offset];
    if version != VERSION {
        return Err(Error::UnsupportedVersion(version));
    }
    offset += 1;

    // Flags
    let flags_byte = data[offset];
    offset += 1;
    let minified = flags_byte & FLAG_MINIFIED != 0;
    let has_checksum = flags_byte & FLAG_CHECKSUM != 0;
    let compression_id = (flags_byte & FLAG_COMPRESSION_MASK) >> FLAG_COMPRESSION_SHIFT;
    let compression = match compression_id {
        1 => CompressionAlgorithm::Gzip,
        _ => CompressionAlgorithm::Deflate,
    };

    // Header length
    let (header_len, n) = decode_varint(data, offset)?;
    offset += n;
    let header_len = header_len as usize;
    let header_end = offset + header_len;

    if header_end > data.len() {
        return Err(Error::TruncatedPacket(
            "Header section extends beyond packet".into(),
        ));
    }

    // Field count
    let (field_count, n) = decode_varint(data, offset)?;
    offset += n;

    // Fields
    let mut fields = Vec::with_capacity(field_count as usize);
    for i in 0..field_count as usize {
        if offset >= header_end {
            return Err(Error::TruncatedPacket(format!(
                "Header field {} extends beyond header section",
                i
            )));
        }

        let field_type = data[offset];
        offset += 1;

        if field_type >= FIELD_CUSTOM_START {
            // Custom field
            if offset >= header_end {
                return Err(Error::TruncatedPacket(format!(
                    "Custom field {} name length missing",
                    i
                )));
            }
            let name_len = data[offset] as usize;
            offset += 1;

            if offset + name_len > header_end {
                return Err(Error::TruncatedPacket(format!(
                    "Custom field {} name truncated",
                    i
                )));
            }
            let name = String::from_utf8_lossy(&data[offset..offset + name_len]).into_owned();
            offset += name_len;

            let (value_len, n) = decode_varint(data, offset)?;
            offset += n;
            let value_len = value_len as usize;

            if offset + value_len > header_end {
                return Err(Error::TruncatedPacket(format!(
                    "Custom field {} value truncated",
                    i
                )));
            }
            let value = String::from_utf8_lossy(&data[offset..offset + value_len]).into_owned();
            offset += value_len;

            fields.push(HeaderField {
                field_type,
                name: Some(name),
                value,
            });
        } else {
            // Known field
            let (value_len, n) = decode_varint(data, offset)?;
            offset += n;
            let value_len = value_len as usize;

            if offset + value_len > header_end {
                return Err(Error::TruncatedPacket(format!(
                    "Field {} (type 0x{:02X}) value truncated",
                    i, field_type
                )));
            }
            let value = String::from_utf8_lossy(&data[offset..offset + value_len]).into_owned();
            offset += value_len;

            fields.push(HeaderField {
                field_type,
                name: None,
                value,
            });
        }
    }

    // Skip remaining header bytes
    offset = header_end;

    // Body section
    if offset >= data.len() {
        return Err(Error::TruncatedPacket("Body section missing".into()));
    }

    let (uncompressed_len, n) = decode_varint(data, offset)?;
    offset += n;

    let crc_size = if has_checksum { CRC32_SIZE } else { 0 };
    let compressed_body_end = data.len() - crc_size;

    if offset > compressed_body_end {
        return Err(Error::TruncatedPacket("Compressed body truncated".into()));
    }

    let compressed_body = &data[offset..compressed_body_end];

    let crc32 = if has_checksum {
        let crc_offset = compressed_body_end;
        if crc_offset + CRC32_SIZE > data.len() {
            return Err(Error::TruncatedPacket("CRC32 checksum truncated".into()));
        }
        Some(
            data[crc_offset] as u32
                | (data[crc_offset + 1] as u32) << 8
                | (data[crc_offset + 2] as u32) << 16
                | (data[crc_offset + 3] as u32) << 24,
        )
    } else {
        None
    };

    Ok((
        version,
        minified,
        has_checksum,
        compression,
        fields,
        compressed_body,
        uncompressed_len as usize,
        crc32,
    ))
}

/// Extract metadata from header fields.
#[allow(clippy::type_complexity)]
fn extract_fields(
    fields: &[HeaderField],
) -> (
    String,
    Option<String>,
    Option<String>,
    Option<String>,
    Option<u64>,
    Option<String>,
    HashMap<String, String>,
) {
    let mut url = String::new();
    let mut etag = None;
    let mut signature = None;
    let mut content_type = None;
    let mut timestamp = None;
    let mut encoding = None;
    let mut custom = HashMap::new();

    for field in fields {
        match field.field_type {
            FIELD_URL => url = field.value.clone(),
            FIELD_ETAG => etag = Some(field.value.clone()),
            FIELD_SIGNATURE => signature = Some(field.value.clone()),
            FIELD_CONTENT_TYPE => content_type = Some(field.value.clone()),
            FIELD_TIMESTAMP => timestamp = field.value.parse().ok(),
            FIELD_ENCODING => encoding = Some(field.value.clone()),
            t if t >= FIELD_CUSTOM_START => {
                if let Some(name) = &field.name {
                    custom.insert(name.clone(), field.value.clone());
                }
            }
            _ => {} // skip unknown known-range fields
        }
    }

    (
        url,
        etag,
        signature,
        content_type,
        timestamp,
        encoding,
        custom,
    )
}

/// Unpack an .hpack binary packet.
pub fn unpack(data: &[u8]) -> Result<UnpackResult, Error> {
    unpack_with_options(data, &UnpackOptions::default())
}

/// Unpack with options.
pub fn unpack_with_options(data: &[u8], options: &UnpackOptions) -> Result<UnpackResult, Error> {
    let (
        version,
        minified,
        _has_checksum,
        compression,
        fields,
        compressed_body,
        _uncompressed_len,
        crc32_value,
    ) = decode_packet(data)?;

    let (url, etag, signature, content_type, timestamp, encoding, custom) = extract_fields(&fields);

    if options.headers_only {
        return Ok(UnpackResult {
            url,
            etag,
            signature,
            content_type,
            timestamp,
            encoding,
            custom,
            html: String::new(),
            version,
            minified,
            compression,
            checksum_valid: None,
        });
    }

    // Decompress
    let decompressed = decompress(compressed_body, compression)?;

    // CRC32 verification
    let checksum_valid = if let Some(expected) = crc32_value {
        let actual = crc32fast::hash(&decompressed);
        if options.verify_checksum && actual != expected {
            return Err(Error::ChecksumMismatch { expected, actual });
        }
        Some(actual == expected)
    } else {
        None
    };

    let html = String::from_utf8(decompressed).map_err(Error::InvalidUtf8)?;

    Ok(UnpackResult {
        url,
        etag,
        signature,
        content_type,
        timestamp,
        encoding,
        custom,
        html,
        version,
        minified,
        compression,
        checksum_valid,
    })
}

/// Read only headers without decompressing the body.
pub fn read_headers(data: &[u8]) -> Result<UnpackResult, Error> {
    unpack_with_options(
        data,
        &UnpackOptions {
            verify_checksum: false,
            headers_only: true,
        },
    )
}

/// Decompress data based on the algorithm.
fn decompress(data: &[u8], algorithm: CompressionAlgorithm) -> Result<Vec<u8>, Error> {
    use std::io::Read;

    match algorithm {
        CompressionAlgorithm::Gzip => {
            let mut decoder = flate2::read::GzDecoder::new(data);
            let mut buf = Vec::new();
            decoder
                .read_to_end(&mut buf)
                .map_err(|e| Error::DecompressionFailed(e.to_string()))?;
            Ok(buf)
        }
        CompressionAlgorithm::Deflate => {
            let mut decoder = flate2::read::DeflateDecoder::new(data);
            let mut buf = Vec::new();
            decoder
                .read_to_end(&mut buf)
                .map_err(|e| Error::DecompressionFailed(e.to_string()))?;
            Ok(buf)
        }
    }
}
